tests/pats.rs

```diff
@@ -0,0 +1,292 @@
+    for i in 0..64 {
+        let scope = Scopes::from_bits_truncate(1 << i);
+        if !Scopes::ALL.contains(scope) {
+            continue;
+        }
+        let req = test::TestRequest::post()
+            .uri("/v2/pat")
+            .append_header(("Authorization", USER_USER_PAT))
+            .set_json(json!({
+                "scopes": scope.bits(),
+                "name": format!("test_pat_scopes Name {}", i),
+                "expires": Utc::now() + Duration::days(1),
+            }))
+            .to_request();
+        let resp = test_env.call(req).await;
+        assert_eq!(
+            resp.status().as_u16(),
+            if scope.restricted() { 400 } else { 200 }
+        );
+    }
+
+    // Create a 'good' PAT for patching
+    let req = test::TestRequest::post()
+        .uri("/v2/pat")
+        .append_header(("Authorization", USER_USER_PAT))
+        .set_json(json!({
+            "scopes": Scopes::COLLECTION_CREATE,
+            "name": "test_pat_scopes Test",
+            "expires": Utc::now() + Duration::days(1),
+        }))
+        .to_request();
+    let resp = test_env.call(req).await;
+    assert_eq!(resp.status().as_u16(), 200);
+    let success: serde_json::Value = test::read_body_json(resp).await;
+    let id = success["id"].as_str().unwrap();
+
+    // Patching to a bad name should fail
+    for name in ["n", "this_name_is_too_long".repeat(16).as_str()] {
+        let req = test::TestRequest::post()
+            .uri("/v2/pat")
+            .append_header(("Authorization", USER_USER_PAT))
+            .set_json(json!({
+                "name": name,
+            }))
+            .to_request();
+        let resp = test_env.call(req).await;
+        assert_eq!(resp.status().as_u16(), 400);
+    }
+
+    // Patching to a bad expiry should fail
+    let req = test::TestRequest::patch()
+        .uri(&format!("/v2/pat/{}", id))
+        .append_header(("Authorization", USER_USER_PAT))
+        .set_json(json!({
+            "expires": Utc::now() - Duration::days(1), // Past
+        }))
+        .to_request();
+    let resp = test_env.call(req).await;
+    assert_eq!(resp.status().as_u16(), 400);
+
+    // Similar to above with PAT creation, patching to a bad scope should fail
+    for i in 0..64 {
+        let scope = Scopes::from_bits_truncate(1 << i);
+        if !Scopes::ALL.contains(scope) {
+            continue;
+        }
+
+        let req = test::TestRequest::patch()
+            .uri(&format!("/v2/pat/{}", id))
+            .append_header(("Authorization", USER_USER_PAT))
+            .set_json(json!({
+                "scopes": scope.bits(),
+            }))
+            .to_request();
+        let resp = test_env.call(req).await;
+        assert_eq!(
+            resp.status().as_u16(),
+            if scope.restricted() { 400 } else { 204 }
+        );
+    }
+
+    // Cleanup test db
+    test_env.cleanup().await;
```

Comment:
It would be great if we could find a way to do this on `Drop` of the test_env, as we're almost guaranteed to forget to call cleanup at some point when writing future tests. The fact that this is async substantially complicates things though. Here are some possible ways to do it that I found: https://stackoverflow.com/a/74511028, https://stackoverflow.com/questions/71541765/rust-async-drop

Reply:
This is on my list for sure, I gave it a go but it was frustrating. We have a similar async drop trait in theseus (for loading bars) but the difference is for that one the runtime continues, whereas the async runtime drops at the end of the test here (and not necessarily cleaning up the database). I tried a couple solutions but they didnt seem to work, so its something on my list to come back to.

Ideally, we drop this manual database creation-dropping system entirely and use #[sqlx::test], but I had similar adjacent problems when I use sqlx::test instead of actix::test.

Reply:
Dang yeah, figured. Something we can try to get in the future.